File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        # check step length
        count = 0
        node = head
        while node and count < k:
            node = node.next
            count += 1
        
        if count < k:
            return head # fewer than k then leave as it is
        
        prev = None
        curr = head
        for _ in range(k):
            nextNode = curr.next
            curr.next = prev
            prev = curr
            curr = nextNode
        
        # recursively handle the rest and attach
        head.next = self.reverseKGroup(curr, k)

        return prev

File: Data Structures & Algorithms/reorder-linked-list/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    void reorderList(ListNode* head) {
        ListNode* prev;
        ListNode* curr = head;
        int count = 0;

        while (head -> next != nullptr) {
            if (count == 0) {
                head = curr;
                curr = curr -> next;
                ++count;
            } else if (count % 2 != 0) {
                prev = curr;
                head -> next = curr -> next; 
                ++count;
            } else {
                head -> next = prev;
                ++count;
            }
        }
        
    }
};
